ihaskell-display/ihaskell-widgets/src/IHaskell/Display/Widgets/Types.hs

```diff
@@ -0,0 +1,249 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE NoMonomorphismRestriction #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE ConstraintKinds #-}
+module IHaskell.Display.Widgets.Types where
```

Comment:
This module seems like the real meat of the implementation.

Can you add some documentation for how this works? Specifically, what are the relevant `data` clauses, type families, and so on? How are singletons used? Basically, if I am someone who has never read this codebase before, understands all the extensions used, and is only marginally familiar with singletons, what do I need to know in order to understand the overall design and the specifics? (Basically a long comment at the top with architecture details and a comment on top of each `data` and `type family` describing their role in the architecture)